Adding docs for EOAP_CWL on CDSE. (#842)
eu-cdse/documentation: APIs/openEO/EOAP-CWL.qmd

---
title: "EOAP CWL"
execute:
  echo: false
jupyter: python3
aliases:
  - /EOAP-CWL.html
---

With openEO, it is now possible to run CWL workflows. On the CDSE backend, this has some extra features, which are described on this page.
Backend agnostic information can be found here: https://github.com/Open-EO/openeo-geopyspark-driver/pull/1548/changes (TODO: Update URL once merged).

## S3 access

CWL code running on this backend will receive short-lived S3 credentials to access the `eodata` bucket.
Those credentials can be accessed through the following environment variables:

- `AWS_ENDPOINT_URL_S3`
- `AWS_ACCESS_KEY_ID`
- `AWS_SECRET_ACCESS_KEY`

They won't work outside the cluster, and are only temporarily valid.
You can use these instead of your own credentials. This way, your Docker images can remain public.

## Docker images

> Only whitelisted Docker images can be used in the cluster. Contact us through the forum or support to get your image prefix whitelisted.

Author: @Slegersj What contacts should I specify to let users contact us for technical topics?

Author: Used: https://helpcenter.dataspace.copernicus.eu/hc/en-gb/requests/new

---
title: "EOAP CWL"
execute:
  echo: false
jupyter: python3
aliases:
  - /EOAP-CWL.html
---

> With openEO, it is now possible to run CWL ([Common Workflow Language](https://www.commonwl.org/)) in the `run_udf` process. This is executed on Calrissian

Reviewer: "This is executed on Calrissian"  looks like incomplete sentence. I'm also not sure this implementation detail must be in inttroduction

Author: Rephrased: "Workflows will be executed using calrissian on kubernetes."

> On the CDSE backend, this has some extra features, which are described on this page.

Reviewer: "extra" compared to what?

Reviewer: I guess you mean extra in addition to what the generic openeo-geopyspark-driver docs list. But then I would swap both sentences.

Also make it more clear that the openeo-geopyspark-driver docs are generic docs: "more information" is pretty vague.

Author: Rephrased. Does it look good like this?

More information can be found here: [udf-eoap-cwl.md](https://github.com/Open-EO/openeo-geopyspark-driver/blob/master/docs/udf-eoap-cwl.md)

## S3 access

CWL workflows running on this backend will receive short-lived S3 credentials with read-only access to the `eodata` bucket.

> Those credentials can be accessed through the following environment variables:

Reviewer: ```suggestion
Those credentials will be available in the following environment variables:
```

with a lot of tools you don't have to actively "access" the variables, just the fact that they are present is often enough to get things working

Author: Using this wording now

- `AWS_ENDPOINT_URL_S3`
- `AWS_ACCESS_KEY_ID`
- `AWS_SECRET_ACCESS_KEY`

They only work inside the cluster environment, and are only temporarily valid.
You can use these instead of your own credentials. This way, your Docker images can remain public.

## Docker images

> Only whitelisted Docker images can be used in the cluster. Contact us through [support](https://helpcenter.dataspace.copernicus.eu/hc/en-gb/requests/new) (You might need to create an account)

Reviewer: maybe be more explicit about what kind of docker images you refer to here. I guess here it's about docker images referenced from CWL workflow documents?

Author: clearified

> As of February 2026, only public Docker images are used.

Reviewer: "public" sounds very open and official (e.g. like there must be an open source organization behind it). 

I think the main requirement is that they can be pulled without need for credentials

Author: Rephrased: "only Docker images that can be pulled without credentials are used"

## Memory limits

> As of February 2026, the memory limit is set to 15GB.

Reviewer: that's the total memory limit for all concurrent containers? or is it per container?

Author: That is not very clear from the documentation:
> When running calrissian, you must provide a limit the the number of CPU cores (--max-cores) and RAM megabytes (--max-ram) to use concurrently

or 

> Maximum amount of RAM to use, e.g 1048576, 512Mi or 2G. Follows k8s resource conventions

or

> - Calrissian requires specifying an envelope of RAM and CPU resources to use in the cluster, so these are provided as `--max-ram` and `--max-cores`

Author: Changed the limit to be on per step basis, instead of being shared over the whole process.
And clarified this wording.

## Debugging locally

To test your CWL workflow locally before running it on the cluster,
you can use [cwltool](https://pypi.org/project/cwltool/) locally.
You might need to provide your own S3 credentials. You can request them
here: https://documentation.dataspace.copernicus.eu/APIs/S3.html

```bash
cwltool \
  --tmpdir-prefix=$HOME/tmp/ \
  --force-docker-pull \
  --leave-container \
  --leave-tmpdir \
  --no-read-only \
  --parallel \
  --preserve-environment=AWS_ENDPOINT_URL_S3 \
  --preserve-environment=AWS_ACCESS_KEY_ID \
  --preserve-environment=AWS_SECRET_ACCESS_KEY \

>   example_workflow.cwl example_parameters.json

Reviewer: It seems a lot of mentioned options are not essential or even risky to use unless you know what you are doing.  I would just have the essentials here and suggest some more advanced options seperately

Author: Removed 2 unneeded parameters. Annoyingly, `--tmpdir-prefix` and `--no-read-only` are needed to be able to run for me.
